frontend/apps/reader/modules/readertoc.lua

```diff
@@ -215,6 +215,34 @@ function ReaderToc:getTocIndexByPage(pn_or_xp)
     return pre_index
 end
 
+function ReaderToc:getAccurateTocIndexByXPointer(xptr)
+    local pageno = self.ui.document:getPageFromXPointer(xptr)
+    local index = self:getTocIndexByPage(pageno) -- get toc entry on the page
+    local i = index
+    if not index or not self.toc[index] then return end
+    local initial_comparison = self.ui.document:compareXPointers(self.toc[index].xpointer, xptr)
+    if initial_comparison and initial_comparison < 0 then
+        while self.toc[i] do
```

Comment:
To avoid a loop (with i=index, that you already checked it's not a candidate), move the `local i = index+/-1` here)